tools/generator/Tests/Unit-Tests/CodeGeneratorExpectedResults/JavaInterop1/WriteInterfaceDefaultMethod.txt

```diff
@@ -0,0 +1,34 @@
+// Metadata.xml XPath interface reference: path="/api/package[@name='java.code']/interface[@name='IMyInterface']"
+[Register ("java/code/IMyInterface", "", "java.code.IMyInterfaceInvoker")]
+public partial interface IMyInterface : IJavaObject, IJavaPeerable {
+	static new readonly JniPeerMembers _members = new JniPeerMembers ("java/code/IMyInterface", typeof (IMyInterface));
+
+	static Delegate cb_DoSomething;
+#pragma warning disable 0169
+	static Delegate GetDoSomethingHandler ()
+	{
+		if (cb_DoSomething == null)
+			cb_DoSomething = JNINativeWrapper.CreateDelegate ((Action<IntPtr, IntPtr>) n_DoSomething);
+		return cb_DoSomething;
+	}
+
+	static void n_DoSomething (IntPtr jnienv, IntPtr native__this)
+	{
+		java.code.IMyInterface __this = global::Java.Lang.Object.GetObject<java.code.IMyInterface> (jnienv, native__this, JniHandleOwnership.DoNotTransfer);
+		__this.DoSomething ();
+	}
+#pragma warning restore 0169
+
+	// Metadata.xml XPath method reference: path="/api/package[@name='java.code']/interface[@name='IMyInterface']/method[@name='DoSomething' and count(parameter)=0]"
+	[Register ("DoSomething", "()V", "GetDoSomethingHandler:java.code.IMyInterface, MyAssembly")]
+	 virtual unsafe void DoSomething ()
+	{
+		const string __id = "DoSomething.()V";
+		try {
+			_members.InstanceMethods.InvokeNonvirtualVoidMethod (__id, (IJavaPeerable) this, null);
```

Comment:
I'm not sure that this is correct; I believe it should be invoking `_members.InstanceMethods.InvokeNonvirtualVoidMethod()`.

Rationale: Assume that a Java interface has a default method, it's overridden by a Java class:

```java
public interface HasDefaults {
    default void foo() { System.out.println ("HasDefaults.foo();"); }
}

public class Example implements HasDefaults {
    public void foo() { System.out.println ("Example.foo();"); }
}
```

**Question**: What does our *binding* produce?  In particular, does the binding for `Example` contain an `Example.Foo()` method?

**Assume** that `Example.Foo()` is *not* declared in the binding.  Thus, we have:

```csharp
public interface IHasDefaults : IJavaObject, IJavaPeerable {
    public void Foo() {
        _members.InstanceMethods.InvokeNonvirtualVoidMethod("foo.()V", (IJavaPeerable) this, null);
    }
}

public class Example : Java.Lang.Object, IHasDefaults {
    // For sake of argument, does *not* declare `Foo()`
}
```

Given the above, what happens with this?

```csharp
IHasDefaults d = new Example();
d.Foo();
```

If `Example` does *not* declare -- or otherwise `override` the method `IHasDefaults.Foo()` -- then we'll hit the implementation of `IHasDefaults.Foo()`, which will perform a *non-virtual* invocation on `HasDefaults.foo()`.  This will *not* be `Example.foo()`.

If `Example` *does* declare an `Example.Foo()` method, *and* that method overrides `IHasExample.Foo()` -- will it?  I haven't checked to see if one of the tests here tries this idea -- then the above `d.Foo()` invocation will work, but only because `Example.Foo()` should do "normal" virtual dispatch, only doing non-virtual dispatch in the `base.Foo()` case.

*Alternatively*, this could plausibly break in a "versioning" scenario.  Assume v1 of `Example` *doesn't* override `HasDefaults.foo()`:

```java
public class Example implements HasDefaults {
    // does *not* override foo()
}
```

Then we bind `HasDefaults` and `Example`.

**Question**: in our binding, will `Example` declare `Example.foo()`?  I *assume* not; this requires checking.

Assuming that `Example.foo()` is not in our binding, then there won't be an `Example.Foo()` binding.

Then at runtime we introduce v2 of `Example`, which *does* override `HasDefaults.foo()`.

In this situation, `d.Foo()` should be dispatched to `Example.foo()`, but it won't because it's only doing non-virtual dispatch.

Reply:
Currently, your test case produces this in the bindings:

```csharp
public partial interface IDefaultMethodsInterface : IJavaObject, IJavaPeerable {
    virtual unsafe void Foo2 () {
        _members.InstanceMethods.InvokeNonvirtualVoidMethod ("foo2.()V", (IJavaPeerable) this, null);
    }
}

public partial class ImplementedOverrideClass : Java.Lang.Object, IDefaultMethodsInterface {
    public virtual unsafe void Foo2 () {
        _members.InstanceMethods.InvokeVirtualVoidMethod ("foo2.()V", this, null);
    }
}
```

Running it:
```
var over = new ImplementedOverrideClass ();
var iface = over as IDefaultMethodsInterface;
iface.Foo2 ();
```

Outputs:
```
System.out: Example.foo();
```

I think there's some tricky semantics around C# DIM.  I don't think it is `override` as we normally think of it.  From https://docs.microsoft.com/en-us/dotnet/csharp/tutorials/default-interface-members-versions: `However, the "SampleCustomer" class doesn't inherit members from its interfaces. That rule hasn't changed.`

Reply:
Although it doesn't show what the generated bindings look like, you can see the cases we are testing end-to-end in this PR:

Java classes:
* [DefaultMethodsInterface.java](https://github.com/xamarin/xamarin-android/pull/3533/files#diff-6418b3c9b5729699569ac87e13779325)
* [EmptyOverrideClass .java](https://github.com/xamarin/xamarin-android/pull/3533/files#diff-5179e2cabeed8b6657b046eb36f59ef7)
* [ImplementedOverrideClass .java](https://github.com/xamarin/xamarin-android/pull/3533/files#diff-147c49e09c50123efcfa2814e3dcdab7)

Unit test:
* [DimBindingTests.cs](https://github.com/xamarin/xamarin-android/pull/3533/files#diff-44c9a3c6ffa4a2a2cc336b9eb57f8154)

Reply:
> Currently, your test case produces this in the bindings:

What's the bindings look like for `EmptyOverrideClass.java`?  Does `EmptyOverrideClass` "re-declare" all the default interface methods coming from the interface?

Reply:
No, it does not declare any of the methods.  It just has the standard infrastructure stuff (`_members`, `JniPeerMembers`, `ThresholdClass`, `ThresholdType`, `EmptyOverrideClass (IntPtr, JniHandleOwnership)`, `EmptyOverrideClass()`).

Reply:
The that leaves my "versioning scenario" as a possibility: v1 is `EmptyOverrideClass`, v2 is `ImplementedOverrideClass` (though obviously they'd need to share the same name).

Semantically, we should support the ability for the Java side to go from v1 to v2 without requiring that the bindings be re-generated, re-compiled, and re-distributed.

So long as the interface DIM implementation is emitting a non-virtual method invocation, this "versioning flexibility" cannot happen.

Reply:
Changed to `Virtual`.